python/lsst/daf/butler/formatters/typeless.py

```diff
@@ -0,0 +1,223 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This software is dual licensed under the GNU General Public License and also
+# under a 3-clause BSD license. Recipients may choose which of these licenses
+# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
+# respectively.  If you choose the GPL option then the following text applies
+# (but note that there is still no warranty even if you opt for BSD instead):
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Support for reading and writing files to a POSIX file system."""
+
+from __future__ import annotations
+
+__all__ = "TypelessFormatter"
+
+import dataclasses
+import logging
+from typing import TYPE_CHECKING, Any
+
+from lsst.daf.butler import FormatterV2
+from lsst.utils.introspection import get_full_type_name
+
+if TYPE_CHECKING:
+    from lsst.daf.butler import StorageClass
+    from lsst.daf.butler.datastore.cache_manager import AbstractDatastoreCacheManager
+
+
+class TypelessFormatter(FormatterV2):
+    """Formatter V2 base class that attempts to coerce generic objects
+    read in subclasses into the correct Python type.
+
+    Notes
+    -----
+    This class provides a ``read()`` method that will run `FormatterV2.read`
+    and coerce the return type using a variety of techniques. Use the
+    standard `FormatterV2` methods for reading bytes/files and writing
+    bytes/files.
```

Comment:
This gets me thinking that we should have different JSON/YAML formatters for different categories of target types (one for things that are natively pydantic models, another for those that have `to_simple` and `from_simple`, etc.).

May be too late for `PropertySet` vs. `TaskMetadata`, of course, but if we'd had different formatters earlier that transition would have been easier to handle, and I'd like to do better in future similar transitions.

Reply:
We could have a PydanticFormatter implementation explicitly. This wouldn't have solved the PropertySet vs TaskMetadata problem though because those are entirely different types and the fix there would have been to have a specialist class from the beginning for task metadata that might have internally used a PropertySet rather than using a generic container class at the start.